File: src/main/java/com/yupi/airouter/controller/InternalChatController.java
package com.yupi.airouter.controller;

import cn.hutool.core.util.StrUtil;
import cn.hutool.extra.servlet.JakartaServletUtil;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.yupi.airouter.annotation.AuthCheck;
import com.yupi.airouter.annotation.RateLimit;
import com.yupi.airouter.common.ResultUtils;
import com.yupi.airouter.constant.UserConstant;
import com.yupi.airouter.exception.BusinessException;
import com.yupi.airouter.exception.ErrorCode;
import com.yupi.airouter.model.dto.chat.ChatMessage;
import com.yupi.airouter.model.dto.chat.ChatRequest;
import com.yupi.airouter.model.dto.chat.ChatResponse;
import com.yupi.airouter.model.entity.User;
import com.yupi.airouter.service.ChatService;
import com.yupi.airouter.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;

/**
 * 内部 Chat 接口（用于前端页面调用）
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@RestController
@RequestMapping("/internal/chat")
@Slf4j
public class InternalChatController {

    @Resource
    private ChatService chatService;

    @Resource
    private UserService userService;

    @Resource
    private ObjectMapper objectMapper;

    /**
     * 内部聊天接口（网页端对话）
     * 支持流式和非流式响应
     */
    @PostMapping(value = "/completions", produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.TEXT_EVENT_STREAM_VALUE})
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "内部聊天接口")
    @RateLimit(type = RateLimit.LimitType.IP, limit = 30)
    public Object chatCompletions(@RequestBody ChatRequest request,
                                  HttpServletRequest httpRequest) {
        User loginUser = userService.getLoginUser(httpRequest);

        // 1. 参数校验
        if (request.getMessages() == null || request.getMessages().isEmpty()) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "messages 不能为空");
        }

        // 2. 获取客户端IP和User-Agent
        String clientIp = JakartaServletUtil.getClientIP(httpRequest);
        String userAgent = httpRequest.getHeader("User-Agent");

        // 3. 判断是否为流式请求
        Boolean stream = request.getStream();
        if (stream != null && stream) {
            // 流式响应（网页端调用，不传 apiKeyId）- 转换为 SSE 格式
            return chatService.chatStream(request, loginUser.getId(), null, clientIp, userAgent)
                    .map(streamResponse -> {
                        try {
                            // 将 StreamResponse 转换为 JSON
                            String json = objectMapper.writeValueAsString(streamResponse);
                            return json + "\n\n";
                        } catch (Exception e) {
                            log.error("Failed to serialize stream response", e);
                            return "";
                        }
                    });
        } else {
            // 非流式响应（网页端调用，不传 apiKeyId）
            ChatResponse response = chatService.chat(request, loginUser.getId(), null, clientIp, userAgent);
            return ResultUtils.success(response);
        }
    }

    /**
     * 内部聊天接口 - 支持文件上传（图片识别、PDF解析）
     * 使用 multipart/form-data 格式
     */
    @PostMapping(value = "/completions/upload", consumes = MediaType.MULTIPART_FORM_DATA_VALUE,
            produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.TEXT_EVENT_STREAM_VALUE})
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "内部聊天接口（支持文件上传）")
    @RateLimit(type = RateLimit.LimitType.IP, limit = 30)
    public Object chatCompletionsWithFile(
            @RequestPart("file") MultipartFile file,
            @RequestParam(value = "model", required = false) String model,
            @RequestParam(value = "messages", required = true) String messagesJson,
            @RequestParam(value = "stream", required = false, defaultValue = "false") Boolean stream,
            @RequestParam(value = "routing_strategy", required = false) String routingStrategy,
            @RequestParam(value = "plugin_key", required = false) String pluginKey,
            @RequestParam(value = "enable_reasoning", required = false) Boolean enableReasoning,
            HttpServletRequest httpRequest) {

        User loginUser = userService.getLoginUser(httpRequest);

        try {
            // 1. 解析消息列表
            List<ChatMessage> messages = objectMapper.readValue(messagesJson, new TypeReference<List<ChatMessage>>() {});
            
            if (messages == null || messages.isEmpty()) {
                throw new BusinessException(ErrorCode.PARAMS_ERROR, "messages 不能为空");
            }

            // 2. 构建 ChatRequest
            ChatRequest request = new ChatRequest();
            request.setMessages(messages);
            request.setModel(model);
            request.setStream(stream);
            request.setRoutingStrategy(routingStrategy);
            request.setPluginKey(pluginKey);
            request.setEnableReasoning(enableReasoning);

            // 3. 处理上传的文件
            if (file != null && !file.isEmpty()) {
                // 直接使用字节数组
                request.setFileBytes(file.getBytes());

                // 设置文件类型
                String contentType = file.getContentType();
                request.setFileType(contentType);

                // 根据文件类型自动设置插件（如果未指定）
                if (StrUtil.isBlank(request.getPluginKey())) {
                    if (contentType != null) {
                        if (contentType.startsWith("image/")) {
                            request.setPluginKey("image_recognition");
                        } else if (contentType.equals("application/pdf")) {
                            request.setPluginKey("pdf_parser");
                        }
                    }
                }

                log.info("文件上传成功: name={}, size={}, type={}, plugin={}",
                        file.getOriginalFilename(), file.getSize(), contentType, request.getPluginKey());
            }

            // 4. 设置默认模型（如果未指定）
            if (request.getModel() == null || request.getModel().isEmpty()) {
                request.setModel("qwen-plus");
            }

            // 5. 获取客户端IP和User-Agent
            String clientIp = JakartaServletUtil.getClientIP(httpRequest);
            String userAgent = httpRequest.getHeader("User-Agent");

            // 6. 判断是否为流式请求
            if (stream != null && stream) {
                return chatService.chatStream(request, loginUser.getId(), null, clientIp, userAgent)
                        .map(streamResponse -> {
                            try {
                                String json = objectMapper.writeValueAsString(streamResponse);
                                return json + "\n\n";
                            } catch (Exception e) {
                                log.error("Failed to serialize stream response", e);
                                return "";
                            }
                        });
            } else {
                ChatResponse response = chatService.chat(request, loginUser.getId(), null, clientIp, userAgent);
                return ResultUtils.success(response);
            }
        } catch (IOException e) {
            log.error("处理请求失败", e);
            throw new BusinessException(ErrorCode.SYSTEM_ERROR, "处理请求失败: " + e.getMessage());
        }
    }
}

File: go-backend/internal/controller/internal_chat_controller.go
package controller

import (
	"encoding/json"
	"io"
	"log"
	"net/http"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"

	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/dto"
	"github.com/yupi/airouter/go-backend/internal/service"
)

type InternalChatController struct {
	chatService *service.ChatService
	userService *service.UserService
}

func NewInternalChatController(chatService *service.ChatService, userService *service.UserService) *InternalChatController {
	return &InternalChatController{
		chatService: chatService,
		userService: userService,
	}
}

func (c *InternalChatController) ChatCompletions(ctx *gin.Context) {
	var request dto.ChatRequest
	if err := ctx.ShouldBindJSON(&request); err != nil {
		log.Printf("internal chat bind request failed: path=%s err=%v", ctx.Request.URL.Path, err)
		common.Error(ctx, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	if len(request.Messages) == 0 {
		common.Error(ctx, errno.ParamsError.Code, "messages 不能为空")
		return
	}
	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		writeServiceError(ctx, err)
		return
	}

	apiKeyID := int64(0)
	clientIP := ctx.ClientIP()
	userAgent := ctx.GetHeader("User-Agent")

	if request.Stream != nil && *request.Stream {
		c.stream(ctx, request, loginUser.ID, apiKeyID, clientIP, userAgent)
		return
	}
	response, err := c.chatService.Chat(request, loginUser.ID, apiKeyID, clientIP, userAgent)
	if err != nil {
		writeServiceError(ctx, err)
		return
	}
	common.Success(ctx, response)
}

func (c *InternalChatController) ChatCompletionsWithFile(ctx *gin.Context) {
	fileHeader, err := ctx.FormFile("file")
	if err != nil {
		log.Printf("internal chat upload missing file: path=%s err=%v", ctx.Request.URL.Path, err)
		common.Error(ctx, errno.ParamsError.Code, "file 不能为空")
		return
	}

	file, err := fileHeader.Open()
	if err != nil {
		log.Printf("internal chat upload open file failed: path=%s err=%v", ctx.Request.URL.Path, err)
		common.Error(ctx, errno.OperationError.Code, "读取上传文件失败")
		return
	}
	defer file.Close()

	fileBytes, readErr := io.ReadAll(file)
	if readErr != nil {
		log.Printf("internal chat upload read file failed: path=%s err=%v", ctx.Request.URL.Path, readErr)
		common.Error(ctx, errno.OperationError.Code, "读取上传文件失败")
		return
	}
	if len(fileBytes) == 0 {
		common.Error(ctx, errno.ParamsError.Code, "上传文件不能为空")
		return
	}

	messagesJSON := strings.TrimSpace(ctx.PostForm("messages"))
	if messagesJSON == "" {
		common.Error(ctx, errno.ParamsError.Code, "messages 不能为空")
		return
	}
	var messages []dto.ChatMessage
	if err = json.Unmarshal([]byte(messagesJSON), &messages); err != nil {
		log.Printf("internal chat upload parse messages failed: raw=%s err=%v", messagesJSON, err)
		common.Error(ctx, errno.ParamsError.Code, "messages 格式错误")
		return
	}
	if len(messages) == 0 {
		common.Error(ctx, errno.ParamsError.Code, "messages 不能为空")
		return
	}

	stream := false
	if rawStream := strings.TrimSpace(ctx.PostForm("stream")); rawStream != "" {
		parsed, parseErr := strconv.ParseBool(rawStream)
		if parseErr != nil {
			log.Printf("internal chat upload parse stream failed: raw=%s err=%v", rawStream, parseErr)
			common.Error(ctx, errno.ParamsError.Code, "stream 参数错误")
			return
		}
		stream = parsed
	}

	var enableReasoning *bool
	if rawReasoning := strings.TrimSpace(ctx.PostForm("enable_reasoning")); rawReasoning != "" {
		parsed, parseErr := strconv.ParseBool(rawReasoning)
		if parseErr != nil {
			log.Printf("internal chat upload parse enable_reasoning failed: raw=%s err=%v", rawReasoning, parseErr)
			common.Error(ctx, errno.ParamsError.Code, "enable_reasoning 参数错误")
			return
		}
		enableReasoning = &parsed
	}

	model := strings.TrimSpace(ctx.PostForm("model"))
	if model == "" {
		model = "qwen-plus"
	}
	pluginKey := strings.TrimSpace(ctx.PostForm("plugin_key"))
	contentType := strings.TrimSpace(fileHeader.Header.Get("Content-Type"))
	if pluginKey == "" {
		if strings.HasPrefix(contentType, "image/") {
			pluginKey = "image_recognition"
		} else if contentType == "application/pdf" {
			pluginKey = "pdf_parser"
		}
	}

	request := dto.ChatRequest{
		Model:           model,
		Messages:        messages,
		Stream:          &stream,
		RoutingStrategy: strings.TrimSpace(ctx.PostForm("routing_strategy")),
		PluginKey:       pluginKey,
		FileBytes:       fileBytes,
		FileType:        contentType,
		EnableReasoning: enableReasoning,
	}

	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		writeServiceError(ctx, err)
		return
	}
	log.Printf("internal chat upload accepted: userId=%d file=%s size=%d type=%s plugin=%s model=%s stream=%t",
		loginUser.ID, fileHeader.Filename, fileHeader.Size, contentType, pluginKey, model, stream)

	apiKeyID := int64(0)
	clientIP := ctx.ClientIP()
	userAgent := ctx.GetHeader("User-Agent")
	if stream {
		c.stream(ctx, request, loginUser.ID, apiKeyID, clientIP, userAgent)
		return
	}
	response, err := c.chatService.Chat(request, loginUser.ID, apiKeyID, clientIP, userAgent)
	if err != nil {
		writeServiceError(ctx, err)
		return
	}
	common.Success(ctx, response)
}

func (c *InternalChatController) stream(ctx *gin.Context, request dto.ChatRequest, userID, apiKeyID int64, clientIP, userAgent string) {
	streamChan, errChan := c.chatService.ChatStream(request, userID, apiKeyID, clientIP, userAgent)
	ctx.Header("Content-Type", "text/event-stream")
	ctx.Header("Cache-Control", "no-cache")
	ctx.Header("Connection", "keep-alive")
	ctx.Status(http.StatusOK)

	flusher, ok := ctx.Writer.(http.Flusher)
	if !ok {
		log.Printf("internal chat stream not supported by writer")
		common.Error(ctx, errno.SystemError.Code, "流式响应不支持")
		return
	}

	for {
		select {
		case chunk, open := <-streamChan:
			if !open {
				return
			}
			payload, marshalErr := json.Marshal(chunk)
			if marshalErr != nil {
				log.Printf("internal chat stream marshal chunk failed: userId=%d apiKeyId=%d err=%v", userID, apiKeyID, marshalErr)
				continue
			}
			_, _ = ctx.Writer.WriteString("data: " + string(payload) + "\n\n")
			flusher.Flush()
		case err, open := <-errChan:
			if open && err != nil {
				log.Printf("internal chat stream error: userId=%d apiKeyId=%d err=%v", userID, apiKeyID, err)
				return
			}
		case <-ctx.Request.Context().Done():
			return
		}
	}
}
